### PR TITLE
vquic: recv_mmsg, use fewer, but larger buffers

### PR DESCRIPTION
**Pull request by icing**
Instead of 64 buffers of 2k size, use recvmmsg with 16 buffers of 64k size each.

refs #15267

### CODE REVIEW

**Comment by icing** [comment]
If we can live with a perm allocation of 1MB per quic connection…otherwise. it‘s not a notable perf issue to malloc. Am 31.10.2024 um 14:16 schrieb Daniel Stenberg ***@***.***>:﻿
@bagder commented on this pull request.



In lib/vquic/vquic.c:
> @@ -443,6 +448,7 @@ static CURLcode recvmmsg_packets(struct Curl_cfilter *cf,
   if(total_nread || result)
     CURL_TRC_CF(data, cf, "recvd %zu packets with %zu bytes -> %d",
                 pkts, total_nread, result);
+  free(bufs);

Shouldn't this allocation be kept for the duration of the transfer to avoid having to repeat it in every invocation of this funtion?

**Comment by icing** [comment]
the alternative would be another temp buffer at the multi…Am 31.10.2024 um 15:02 schrieb Stefan Eissing ***@***.***>:﻿If we can live with a perm allocation of 1MB per quic connection…otherwise. it‘s not a notable perf issue to malloc. Am 31.10.2024 um 14:16 schrieb Daniel Stenberg ***@***.***>:﻿
@bagder commented on this pull request.



In lib/vquic/vquic.c:
> @@ -443,6 +448,7 @@ static CURLcode recvmmsg_packets(struct Curl_cfilter *cf,
   if(total_nread || result)
     CURL_TRC_CF(data, cf, "recvd %zu packets with %zu bytes -> %d",
                 pkts, total_nread, result);
+  free(bufs);

Shouldn't this allocation be kept for the duration of the transfer to avoid having to repeat it in every invocation of this funtion?

**Comment by bagder** [comment]
> the alternative would be another temp buffer at the multi

Right, since it is only used one at a time anyway within the multi. I prefer that - since we need to use the memory anyway, keeping it allocated is more efficient.

File: lib/vquic/vquic.c [hunk 1]
```diff
@@ -385,7 +390,7 @@ static CURLcode recvmmsg_packets(struct Curl_cfilter *cf,
       mmsg[i].msg_hdr.msg_name = &remote_addr[i];
       mmsg[i].msg_hdr.msg_namelen = sizeof(remote_addr[i]);
       mmsg[i].msg_hdr.msg_control = &msg_ctrl[i];
```
Comment (tatsuhiro-t) [inline lib/vquic/vquic.c:392]: The offset needs to be adjusted with CMSG_SPACE boundary:

```suggestion
      mmsg[i].msg_hdr.msg_control = &msg_ctrl[i * CMSG_SPACE(sizeof(int))];
```

Comment (icing) [reply lib/vquic/vquic.c]: Thanks, just pushed this change.

**Comment by icing** [comment]
> > the alternative would be another temp buffer at the multi
> 
> Right, since it is only used one at a time anyway within the multi. I prefer that - since we need to use the memory anyway, keeping it allocated is more efficient.

Added in the last commit.

**Review by bagder — APPROVED** [review]

**Comment by koujaz** [comment]
Tested commit 9bb211e7f on our setup and can confirm the fix works.
There is no packet loss induced by recvmmsg.

**Comment by bagder** [comment]
Thanks for verifying!

**Pull request closed without merging**